src/NServiceBus.Gateway/Channels/Http/HttpChannelReceiver.cs

```diff
@@ -49,20 +48,67 @@ public void Dispose()
             {
                 listener.Close();
             }
-            if (scheduler != null)
+            try
+            {
+                receiverTask.Wait(TimeSpan.FromSeconds(10));
+            }
+            catch (AggregateException ex)
+            {
+                ex.Handle(e => e is TaskCanceledException);
+            }
+
+            // Do not dispose the task, see http://blogs.msdn.com/b/pfxteam/archive/2012/03/25/10287435.aspx
+            if (concurencyLimiter != null)
+            {
+                concurencyLimiter.Dispose();
+            }
+        }
+
+        private async Task HandleRequestAsync(object o)
+        {
+            var cancellationToken = (CancellationToken)o;
+
+            try
+            {
+                await concurencyLimiter.WaitAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
This concurrency check is being done outside the processing loop. Is that right?